src/main/java/com/google/devtools/build/lib/syntax/SkylarkList.java

```diff
@@ -201,6 +206,312 @@ public int hashCode() {
+   * list} is returned.
+   */
+  @SkylarkModule(
+      name = "range",
+      category = SkylarkModuleCategory.BUILTIN,
+      doc =
+          "A language built-in type to support ranges. Example of range literal:<br>"
+              + "<pre class=language-python>x = range(1, 10, 3)</pre>"
+              + "Accessing elements is possible using indexing (starts from <code>0</code>):<br>"
+              + "<pre class=language-python>e = x[1]   # e == 2</pre>"
+              + "Ranges do not support the <code>+</code> operator for concatenation."
+              + "Similar to strings, ranges support slice operations:"
+              + "<pre class=language-python>range(10)[1:3]   # range(1, 3)\n"
+              + "range(10)[::2]  # range(0, 10, 2)\n"
+              + "range(10)[3:0:-1]  # range(3, 0, -1)</pre>"
+              + "Ranges are immutable, as in Python 3.")
+  public static final class RangeList extends SkylarkList<Integer> {
+
+    private final int step;
+
+    /** Provides access to range elements based on their index. */
+    private static class RangeListView extends AbstractList<Integer> {
+
+      /** Iterator for increasing sequences. */
+      private static class IncreasingIterator extends UnmodifiableIterator<Integer> {
+        private final int stop;
+        private final int step;
+
+        private int cursor;
+
+        private IncreasingIterator(int start, int stop, int step) {
+          this.cursor = start;
+          this.stop = stop;
+          this.step = step;
+        }
+
+        @Override
+        public boolean hasNext() {
+          return cursor < stop;
+        }
+
+        @Override
+        public Integer next() {
+          if (!hasNext()) {
+            throw new NoSuchElementException();
+          }
+          int current = cursor;
+          cursor += step;
+          return current;
+        }
+      }
+
+      /** Iterator for decreasing sequences. */
+      private static class DecreasingIterator extends UnmodifiableIterator<Integer> {
+        private final int stop;
+        private final int step;
+
+        private int cursor;
+
+        private DecreasingIterator(int start, int stop, int step) {
+          this.cursor = start;
+          this.stop = stop;
+          this.step = step;
+        }
+
+        @Override
+        public boolean hasNext() {
+          return cursor > stop;
+        }
+
+        @Override
+        public Integer next() {
+          if (!hasNext()) {
+            throw new NoSuchElementException();
+          }
+          int current = cursor;
+          cursor += step;
+          return current;
+        }
+      }
+
+      /**
+       * @return The size of the range specified by {@code start}, {@code stop} and {@code step}.
+       * Python version: https://github.com/python/cpython/blob/09bb918a61031377d720f1a0fa1fe53c962791b6/Objects/rangeobject.c#L144
+       */
+      private static int computeSize(int start, int stop, int step) {
+        int low;
```

Comment:
In light of the inclusion/exclusion confusion elsewhere, can you add a comment indicating that this is a half-open interval and it doesn't matter which var represents the included endpoint?